# JUMPlus Python Project 4 - Part 1

## Video Game Sales Basic Data Cleaning

### by Nicholas Crossman

In this project, we've been given a dataset with some missing values. 

Collecting data in the real world is often messy, and some records will be incomplete. It's essential as a data engineer to know how to deal with these missing values. 

There are more advanced techniques like imputing missing values, which tries to fill in missing columns using machine learning. However, that's beyond the scope of this practice project. In this case, we're just going to try and understand which values are missing, and which columns we can keep and which we should ignore.

First, we read in the data from the `.csv` file.

In [16]:
import pandas as pd
import matplotlib as plt

data = pd.read_csv("Video_Games_Sales_as_at_22_Dec_2016.csv")

Let's see the first 5 entries to get an idea of what we're dealing with.

In [17]:
data.head()

Name Platform  Year_of_Release         Genre Publisher  \
0                Wii Sports      Wii           2006.0        Sports  Nintendo   
1         Super Mario Bros.      NES           1985.0      Platform  Nintendo   
2            Mario Kart Wii      Wii           2008.0        Racing  Nintendo   
3         Wii Sports Resort      Wii           2009.0        Sports  Nintendo   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  Critic_Score  \
0     41.36     28.96      3.77         8.45         82.53          76.0   
1     29.08      3.58      6.81         0.77         40.24           NaN   
2     15.68     12.76      3.79         3.29         35.52          82.0   
3     15.61     10.93      3.28         2.95         32.77          80.0   
4     11.27      8.89     10.22         1.00         31.37           NaN   

   Critic_Count  User_Score  User_Count Developer Rating  
0          51.0         8.0       322.0  Nintendo      E  
1           NaN         NaN         NaN       NaN    NaN  
2          73.0         8.3       709.0  Nintendo      E  
3          73.0         8.0       192.0  Nintendo      E  
4           NaN         NaN         NaN       NaN    NaN

In [18]:
data.shape

(16719, 16)

The `shape` method tells us there are 16 columns in total.

We can already see some `NaN` values indicating missing data. Since this is just the first 5 entries, we need a query to see which columns have null or NaN values.

Helpfully, Pandas provides the `isna()` method, which finds `null` or `NaN` values.

In [19]:
data.columns[data.isna().any()].tolist()

['Name',
 'Year_of_Release',
 'Genre',
 'Publisher',
 'Critic_Score',
 'Critic_Count',
 'User_Score',
 'User_Count',
 'Developer',
 'Rating']

That's 10 columns out of 16 with some missing values. 

We should try and get a count of missing values in each column, because this is too many to throw out all of them. We need to choose which ones are useful and which have too much missing data.

Let's try and count how many null values are in each of these columns. Remember that there are 16,719 records in total.

In [20]:
data.isna().sum()

Name                  2
Platform              0
Year_of_Release     269
Genre                 2
Publisher            54
NA_Sales              0
EU_Sales              0
JP_Sales              0
Other_Sales           0
Global_Sales          0
Critic_Score       8582
Critic_Count       8582
User_Score         9129
User_Count         9129
Developer          6623
Rating             6769
dtype: int64

In this case, lower numbers mean fewer missing values. The sales data seems to have no missing data, which is encouraging. However, the critic scores and user scores seem to be missing a lot of values. Let's find the percent of how many records are missing. We'll use `data.shape[0]` to get the number of records in the data set.

In [21]:
( data.isna().sum() / data.shape[0] ) * 100

Name                0.011962
Platform            0.000000
Year_of_Release     1.608948
Genre               0.011962
Publisher           0.322986
NA_Sales            0.000000
EU_Sales            0.000000
JP_Sales            0.000000
Other_Sales         0.000000
Global_Sales        0.000000
Critic_Score       51.330821
Critic_Count       51.330821
User_Score         54.602548
User_Count         54.602548
Developer          39.613613
Rating             40.486871
dtype: float64

Now we can see just how unreliable the last 6 columns are. They're missing between 39% and 55% of their values, which makes them useless for most of our analysis.

We will use the `.drop()` function to drop columns with more than 6000 missing values, which will drop the last 6 columns.

In [23]:
# drops all columns with over 6000 missing values. The inplace flag makes it change the value of `data` instead of needing to reassign it
data.drop( data.columns[data.apply(lambda col: col.isna().sum() > 6000)] , axis=1, inplace=True)

data.head()

Name Platform  Year_of_Release         Genre Publisher  \
0                Wii Sports      Wii           2006.0        Sports  Nintendo   
1         Super Mario Bros.      NES           1985.0      Platform  Nintendo   
2            Mario Kart Wii      Wii           2008.0        Racing  Nintendo   
3         Wii Sports Resort      Wii           2009.0        Sports  Nintendo   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  
0     41.36     28.96      3.77         8.45         82.53  
1     29.08      3.58      6.81         0.77         40.24  
2     15.68     12.76      3.79         3.29         35.52  
3     15.61     10.93      3.28         2.95         32.77  
4     11.27      8.89     10.22         1.00         31.37

Now our columns are much more reliable. Let's check our null values again.

In [24]:
data.isna().sum()

Name                 2
Platform             0
Year_of_Release    269
Genre                2
Publisher           54
NA_Sales             0
EU_Sales             0
JP_Sales             0
Other_Sales          0
Global_Sales         0
dtype: int64

That's much more reasonable in a data set of over 16,000 entries. To make sure we don't encounter any errors, we can drop every row that contains a null value.

In [28]:
data.dropna(inplace=True)

Now our data has been cleaned. We should export the cleaned dataset as a `.csv` so we can perform our analysis in another notebook. We should include the `index=False` flag to prevent the DataFrame from saving the index of each row as a new column, which would make things more complicated.

In [29]:
data.to_csv('cleaned_game_sales.csv', index=False)

Let's quickly read in the file and check the first few entries to see if it was saved correctly.

In [30]:
saved_data = pd.read_csv('cleaned_game_sales.csv')
saved_data.head()

Name Platform  Year_of_Release         Genre Publisher  \
0                Wii Sports      Wii           2006.0        Sports  Nintendo   
1         Super Mario Bros.      NES           1985.0      Platform  Nintendo   
2            Mario Kart Wii      Wii           2008.0        Racing  Nintendo   
3         Wii Sports Resort      Wii           2009.0        Sports  Nintendo   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  
0     41.36     28.96      3.77         8.45         82.53  
1     29.08      3.58      6.81         0.77         40.24  
2     15.68     12.76      3.79         3.29         35.52  
3     15.61     10.93      3.28         2.95         32.77  
4     11.27      8.89     10.22         1.00         31.37

## Conclusion

We've performed some simple data cleaning to turn what would have been a problematic dataset into a much better one. 

It's important to drop columns with a large amount of missing information, because they make analysis or prediction very unreliable. If we hadn't dropped the columns before removing all rows with null data, at least half of our data set would have been thrown out.

The next step is to perform some analysis, which we'll do in the next notebook.